hack/lib/deploy-kubeadmiral.sh

```diff
@@ -0,0 +1,202 @@
+    local context_name=$2
+    local pod_label=$3
+    local pod_namespace=$4
+
+    echo "wait the ${pod_label} ready..."
+    set +e
+    deploy::kubectl_with_retry --kubeconfig="${kubeconfig_path}" --context="${context_name}" wait --for=condition=Ready --timeout=30s pods -l ${pod_label} -n ${pod_namespace}
+    ret=$?
+    set -e
+    if [ $ret -ne 0 ];then
+      echo "kubectl describe info:"
+      kubectl --kubeconfig="${kubeconfig_path}" --context="${context_name}" describe pod -l ${pod_label} -n ${pod_namespace}
+      echo "kubectl logs info:"
+      kubectl --kubeconfig="${kubeconfig_path}" --context="${context_name}" logs -l ${pod_label} -n ${pod_namespace}
+    fi
+    return ${ret}
+}
+
+# deploy::create_signing_certkey creates a CA, args are sudo, dest-dir, ca-id, cn, purpose
+function deploy::create_signing_certkey {
+    local sudo=$1
+    local dest_dir=$2
+    local id=$3
+    local cn=$4
+    local purpose=$5
+    OPENSSL_BIN=$(command -v openssl)
+    # Create ca
+    ${sudo} /usr/bin/env bash -e <<EOF
+    rm -f "${dest_dir}/${id}.crt" "${dest_dir}/${id}.key"
+    ${OPENSSL_BIN} req -nodes \
+      -newkey rsa:2048 -keyout "${dest_dir}/${id}.key" -out "${dest_dir}/${id}.crt" \
+      -x509 -sha256 -new -days 3650 -subj "/CN=${cn}/"
+    echo '{"signing":{"default":{"expiry":"43800h","usages":["signing","key encipherment",${purpose}]}}}' > "${dest_dir}/${id}-config.json"
+EOF
+}
+
+# deploy::create_certkey signs a certificate: args are sudo, dest-dir, ca, filename (roughly), subject, hosts...
+function deploy::create_certkey {
+    local sudo=$1
+    local dest_dir=$2
+    local ca=$3
+    local id=$4
+    local cn=${5:-$4}
+    local og=$6
+    local hosts=""
+    local SEP=""
+    shift 6
+    while [[ -n "${1:-}" ]]; do
+        hosts+="${SEP}\"$1\""
+        SEP=","
+        shift 1
+    done
+    ${sudo} /usr/bin/env bash -e <<EOF
+    cd ${dest_dir}
+    echo '{"CN":"${cn}","hosts":[${hosts}],"names":[{"O":"${og}"}],"key":{"algo":"rsa","size":2048}}' \
+      | ${CFSSL_BIN} gencert -ca=${ca}.crt -ca-key=${ca}.key -config=${ca}-config.json - \
+      | ${CFSSLJSON_BIN} -bare ${id}
+    mv "${id}-key.pem" "${id}.key"
+    mv "${id}.pem" "${id}.crt"
+    rm -f "${id}.csr"
+EOF
+}
+
+# generate a secret to store the certificates
+function deploy::generate_cert_secret {
+  local root_ca_file_path=$1
+  local root_ca_key_path=$2
+  local cert_yaml_path=$3
+  local kubeconfig_path=$4
+  local meta_cluster_context=$5
+
+  local kubeadmiral_ca=$(base64 -i "${root_ca_file_path}" | tr -d '\r\n')
+  local kubeadmiral_ca_key=$(base64 -i "${root_ca_key_path}" | tr -d '\r\n')
+
+  local TEMP_PATH=$(mktemp -d)
+
+  cp -rf "${cert_yaml_path}"/kubeadmiral-cert-secret.yaml "${TEMP_PATH}"/kubeadmiral-cert-secret-tmp.yaml
+  cp -rf "${cert_yaml_path}"/kubeconfig-secret.yaml "${TEMP_PATH}"/kubeconfig-secret-tmp.yaml
+
+  sed -i'' -e "s/{{ca_crt}}/${kubeadmiral_ca}/g" "${TEMP_PATH}"/kubeadmiral-cert-secret-tmp.yaml
```

Comment:
can we clean up this function and reuse the same replace commands for all files? furthermore, the temp path seems unnecessary since we can just pipe the `sed` output to `kubctl apply -f -` directly?

Reply:
> can we clean up this function and reuse the same replace commands for all files? furthermore, the temp path seems unnecessary since we can just pipe the `sed` output to `kubctl apply -f -` directly?

This function has been cleaned up, please take a review.